### PR TITLE
Ehancement/5221 sharing modal ga events

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue:

- #5221

## Relevant technical choices

* GA events for the DashboardSharingSettings component will be implemented in #4822 
* Updated the haveSharingSettingsChanged selector to correctly detect changes when keys are used.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/googlesitekit/modules/datastore/sharing-settings.js [hunk 1]
```diff
@@ -444,18 +447,27 @@ const baseSelectors = {
...
+			if ( Array.isArray( savedKeys ) && savedKeys.length === 1 ) {
+				savedKeys = savedKeys[ 0 ];
+			}
+
+			if ( Array.isArray( originalKeys ) && Array.isArray( savedKeys ) ) {
+				originalKeys.sort();
+				savedKeys.sort();
+			}
+
+			return ! isEqual( originalKeys, savedKeys );
```
Comment (tofumatt) [inline assets/js/googlesitekit/modules/datastore/sharing-settings.js:465]: What's the reason behind extracting the values in a one-item-array? It seems to add complexity to the code when something like: `isEqual( [ 'a' ], [ 'a' ] )` should work the same as `isEqual( 'a', 'a' )`, which is all this is doing.

What about simplifying this to:

```suggestion
			const originalKeys = at( sharingSettings, keys );
			const savedKeys = at( savedSharingSettings, keys );
			
			originalKeys.sort();
			savedKeys.sort();

			return ! isEqual( originalKeys, savedKeys );
```

File: assets/js/components/dashboard-sharing/UserRoleSelect.js [hunk 2]
```diff
@@ -56,12 +60,25 @@ export default function UserRoleSelect( { moduleSlug } ) {
...
+			trackEvent(
+				`${ viewContext }_sharing`,
+				'change_shared_roles',
+				moduleSlug
+			);
+		}
+
 		setEditMode( ! editMode );
-	}, [ editMode, setEditMode ] );
+	}, [ editMode, setEditMode, rolesChanged, viewContext, moduleSlug ] );
```
Comment (tofumatt) [inline assets/js/components/dashboard-sharing/UserRoleSelect.js:81]: Looks like this needs the `rolesChanged` variable removed:
```suggestion
	}, [ editMode, setEditMode, viewContext, moduleSlug ] );
```

**Review by tofumatt — DISMISSED** [review]
Looks good, but I have a question about the change to `haveSharingSettingsChanged`.

File: assets/js/components/dashboard-sharing/UserRoleSelect.js (hunk 2 above)
Comment (techanvil) [reply assets/js/components/dashboard-sharing/UserRoleSelect.js]: Hi @tofumatt, having taken a look into this I realised the changes didn't quite meet the spec. I have updated the code so it only tracks an event when the user has made changes to the roles between the point of entering edit mode and then leaving it. As a result, this dependency list has also been updated.

File: assets/js/googlesitekit/modules/datastore/sharing-settings.js (hunk 1 above)
Comment (techanvil) [reply assets/js/googlesitekit/modules/datastore/sharing-settings.js]: I've reverted this change, as `haveSharingSettingsChanged` is no longer being called with the `keys` argument, and it's not clear that we'd need this modification. The existing code which uses Lodash `pick`/`isEqual` seems in keeping with the non-keys version, and we can revisit this in future, if necessary, when we do start using `haveSharingSettingsChanged` with `keys`.

**Review by aaemnnosttv — APPROVED** [review]
LGTM 👍

**Pull request merged**